Sources/SwiftRefactor/ExpandEditorPlaceholder.swift

```diff
@@ -0,0 +1,346 @@
+/// ```
+///   'T##' display-string '##' type-string ('##' type-for-expansion-string)?
+///   'T##' display-and-type-string
+///   display-string
+/// ```
+/// It is required that '##' is not a valid substring of display-string or
+/// type-string. If this ends up not the case for some reason, we can consider
+/// adding escaping for '##'.
+///
+/// The type string provided in the placeholder (preferring
+/// `type-for-expansion-string`), is parsed into a syntax node. If that node is
+/// a `FunctionTypeSyntax` then the placeholder is expanded into a
+/// `ClosureExprSyntax`. Otherwise it is expanded as is, which is also the case
+/// for when only a display string is provided.
+///
+/// ## Function Typed Placeholder
+/// ### Before
+/// ```swift
+/// <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>
+/// ```
+///
+/// ### After
+/// ```swift
+/// { someInt in
+///   <#T##String#>
+/// }
+/// ```
+///
+/// ## Other Type Placeholder
+/// ### Before
+/// ```swift
+/// <#T##Int#>
+/// ```
+///
+/// ### After
+/// ```swift
+/// Int
+/// ```
+///
+/// ## No Type Placeholder
+/// ### Before
+/// ```swift
+/// <#anything here#>
+/// ```
+///
+/// ## After
+/// ```swift
+/// anything here
+/// ```
+public struct ExpandEditorPlaceholder: EditRefactoringProvider {
+  public static func isPlaceholder(_ str: String) -> Bool {
+    return str.hasPrefix(placeholderStart) && str.hasSuffix(placeholderEnd)
+  }
+
+  public static func wrapInPlaceholder(_ str: String) -> String {
+    return placeholderStart + str + placeholderEnd
+  }
+
+  public static func wrapInTypePlaceholder(_ str: String, type: String) -> String {
+    return Self.wrapInPlaceholder("T##" + str + "##" + type)
+  }
+
+  public static func textRefactor(syntax token: TokenSyntax, in context: Void) -> [Edit] {
+    guard let placeholder = EditorPlaceholderData(token: token) else {
+      return []
+    }
+
+    let expanded: String
+    switch placeholder {
+    case let .basic(text):
+      expanded = String(text)
+    case let .typed(text, type):
+      if let functionType = type.as(FunctionTypeSyntax.self) {
+        expanded = functionType.closureExpansion.formatted().description
+      } else {
+        expanded = String(text)
+      }
+    }
+
+    return [Edit.replace(token, with: token.leadingTrivia.description + expanded + token.trailingTrivia.description)]
+  }
+}
+
+/// If a function-typed placeholder is the argument to a non-trailing closure
+/// call, expands it and any adjacent function-typed placeholders to trailing
+/// closures on that call. All other placeholders will expand as per
+/// `ExpandEditorPlaceholder`.
+///
+/// ## Before
+/// ```swift
+/// foo(closure1: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, normalArg: <#T##Int#>, closure2: { ... }, closure3: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, closure4: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>
```

Comment:
This is pretty awful to read. Could you put the parameters onto different lines?

Reply:
Heh, I didn't notice because they're placeholders in the editor 😅, but sure.